Source/Pawnmorphs/Esoteria/Genebank/ChamberDatabase.cs

```diff
@@ -0,0 +1,423 @@
+		/// <param name="reason">The reason if it didn't get added.</param>
+		/// <returns>
+		///     <c>true</c> if the provided value was added to the genebank; otherwise, <c>false</c>.
+		/// </returns>
+		public bool TryAddToDatabase<T>([NotNull] GenebankEntry<T> entry, out string reason)
+        {
+            if (CanAddToDatabase(entry, out reason) == false)
+                return false;
+
+            return AddToDatabase(entry);
+        }
+
+        /// <summary>
+        /// Determines whether this the provided value can be added to the genebank.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="entry">The entry.</param>
+        /// <returns>
+        ///   <c>true</c> if the value can be added to teh genebank; otherwise, <c>false</c>.
+        /// </returns>
+        public bool CanAddToDatabase<T>([NotNull] GenebankEntry<T> entry)
+		{
+			return CanAddToDatabase(entry, out _);
+		}
+
+		/// <summary>
+		///     Determines whether this instance with the specified mutation definition can be added to the database
+		/// </summary>
+		/// <param name="entry">The genebank entry to check.</param>
+		/// <param name="reason">The reason.</param>
+		/// <returns>
+		///     <c>true</c> if this instance with the specified mutation definition  [can add to database]  otherwise, <c>false</c>
+		///     .
+		/// </returns>
+		public bool CanAddToDatabase<T>([NotNull] GenebankEntry<T> entry, out string reason)
+		{
+			if (entry == null) 
+                throw new ArgumentNullException(nameof(entry));
+
+            Type entryType = typeof(GenebankEntry<T>);
+			if (_genebankDatabase.ContainsKey(entryType) && _genebankDatabase[entryType].Contains(entry))
+			{
+				reason = ALREADY_TAGGED_REASON.Translate(entry.GetCaption());
+				return false;
+			}
+
+            int requiredStorage = entry.GetRequiredStorage();
+			if (FreeStorage < requiredStorage)
+			{
+				reason = NOT_ENOUGH_STORAGE_REASON.Translate(entry.GetCaption(), DatabaseUtilities.GetStorageString(requiredStorage), DatabaseUtilities.GetStorageString(FreeStorage));
+				return false;
+			}
+
+			if (!CanTag)
+			{
+				reason = NOT_ENOUGH_POWER.Translate();
+				return false;
+			}
+
+            if (entry.CanAddToDatabase(this, out reason) == false)
+            {
+                return false;
+            }
+
+			reason = "";
+			return true;
+		}
+
+
+
+		/// <summary>
+		///     Determines whether any of the specified mutation definitions can be
+		///     added to the database, and outputs an error if not.
+		/// </summary>
+		/// <param name="mutationDefs">The mutation definitions.</param>
+		/// <param name="reason">The reason the mutation cannot be ad.</param>
+		/// <returns>
+		///     <c>true</c> if at least one mutation definition can be added to database, otherwise <c>false</c>.
+		/// </returns>
+		public bool CanAddAnyToDatabase([NotNull] IEnumerable<MutationDef> mutationDefs, out string reason)
```

Comment:
This method could be changed to accept IEnumberable<GenebankEntry<T>> instead.  I don't know if it's really needed for anything besides multiple mutations at the moment, but it's futureproofing if nothing else.

Reply:
I am open to suggestions!
I tried to think of a way, but this was the only way i could think of to what wrapper type (logic implementation) to use to check against.
And allow instantiating once to reuse.

I questioned if we even needed this at all, but it does logic checks 1 step for each items at a time and then returns specialised error messages.

Reply:
Hrm... after thinking about it, you *could* generalize it, but as you mentioned before you'd end up creating a lot of garbage objects no matter what.  I checked the usages, and it's only used right now to provide feedback for trying to download a genome.

It's probably not worth generalizing yet after all.  But I did have a thought on how to avoid making GenebankEntry<T> mutable (I'll drop another comment on the particular line)